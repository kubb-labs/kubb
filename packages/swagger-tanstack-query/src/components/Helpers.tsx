import path from 'node:path'
import url from 'node:url'

import { File } from '@kubb/react'

import type { ReactNode } from 'react'

export const templatesPath = path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), '../templates')

const typesPath = path.resolve(templatesPath, './types.ts')

export function Helpers(): ReactNode {
  return (
    <>
      <File.Source path={typesPath} print removeComments></File.Source>
    </>
  )
}
