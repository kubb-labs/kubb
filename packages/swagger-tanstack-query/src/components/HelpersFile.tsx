import path from 'node:path'
import url from 'node:url'

import { File } from '@kubb/react'

import type { ReactNode } from 'react'

export const templatesPath = path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), '../templates')

const typesPath = path.resolve(templatesPath, './types.ts')

type Props = {
  id?: string
  path: string
}

export function HelpersFile({ id, path: resolvedPath }: Props): ReactNode {
  return (
    <File id={id} override baseName={'types.ts'} path={resolvedPath}>
      <File.Source path={typesPath} print removeComments />
    </File>
  )
}
