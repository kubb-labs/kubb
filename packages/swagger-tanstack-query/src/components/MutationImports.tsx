import { PackageManager } from '@kubb/core'
import { File } from '@kubb/react'

import { getImportNames } from '../utils.ts'

import type { ReactNode } from 'react'

type TemplateProps = {
  path: string
  optionsType: string
  resultType: string
  hookName: string
}

function Template({ path, hookName, optionsType, resultType }: TemplateProps): ReactNode {
  return (
    <>
      <File.Import name={[optionsType, resultType]} path={path} isTypeOnly />
      <File.Import name={[hookName]} path={path} />
    </>
  )
}

type FrameworkProps = Partial<TemplateProps>

const defaultTemplates = {
  get react() {
    return function ({ ...rest }: FrameworkProps): ReactNode {
      const importNames = getImportNames()

      return <Template {...importNames.mutation.react} {...rest} />
    }
  },
  get solid() {
    return function ({ ...rest }: FrameworkProps): ReactNode {
      const importNames = getImportNames()

      return <Template {...importNames.mutation.solid} {...rest} />
    }
  },
  get svelte() {
    return function ({ ...rest }: FrameworkProps): ReactNode {
      const importNames = getImportNames()

      return <Template {...importNames.mutation.svelte} {...rest} />
    }
  },
  get vue() {
    return function ({ ...rest }: FrameworkProps): ReactNode {
      const importNames = getImportNames()
      const isV5 = new PackageManager().isValidSync(/@tanstack/, '>=5')
      const path = '@tanstack/vue-query'

      return (
        <>
          {isV5 && <Template {...importNames.mutation.vue} {...rest} />}

          {!isV5 && (
            <>
              <File.Import name={[importNames.mutation.vue.resultType]} path={path} isTypeOnly />
              <File.Import name={[importNames.mutation.vue.optionsType]} path={'@tanstack/vue-query/build/lib/useMutation'} isTypeOnly />
              <File.Import name={[importNames.mutation.vue.hookName]} path={path} />
            </>
          )}
          <File.Import name={['unref']} path={'vue'} />
          <File.Import name={['MaybeRef']} path={'vue'} isTypeOnly />
        </>
      )
    }
  },
} as const

type Props = {
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<FrameworkProps>
}

export function MutationImports({ Template = defaultTemplates.react }: Props): ReactNode {
  return <Template />
}

MutationImports.templates = defaultTemplates
