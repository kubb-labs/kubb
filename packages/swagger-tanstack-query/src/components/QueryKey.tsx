import { FunctionParams, URLPath } from '@kubb/core/utils'
import { Function, Type, usePlugin } from '@kubb/react'
import { useOperation, useOperationSchemas } from '@kubb/swagger/hooks'
import { getASTParams, isRequired } from '@kubb/swagger/utils'

import type { ReactNode } from 'react'
import type { PluginOptions } from '../types'

type TemplateProps = {
  /**
   * Name of the function
   */
  name: string
  /**
   * TypeName of the function in PascalCase
   */
  typeName: string
  /**
   * Parameters/options/props that need to be used
   */
  params: string
  /**
   * Generics that needs to be added for TypeScript
   */
  generics?: string
  /**
   * ReturnType(see async for adding Promise type)
   */
  returnType?: string
  /**
   * Options for JSdocs
   */
  JSDoc?: {
    comments: string[]
  }
  keys?: string
}

function Template({ name, typeName, params, generics, returnType, JSDoc, keys }: TemplateProps): ReactNode {
  return (
    <>
      <Function.Arrow name={name} export generics={generics} params={params} returnType={returnType} singleLine JSDoc={JSDoc}>
        {`[${keys}] as const`}
      </Function.Arrow>

      <Type name={typeName} export>
        {`ReturnType<typeof ${name}>`}
      </Type>
    </>
  )
}

type FrameworkProps = TemplateProps & {
  context: {
    factory: {
      name: string
    }
  }
}

const defaultTemplates = {
  get react() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get solid() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get svelte() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get vue() {
    return function ({ context, ...rest }: FrameworkProps): ReactNode {
      const { factory } = context

      const {
        options: { pathParamsType },
      } = usePlugin<PluginOptions>()
      const schemas = useOperationSchemas()
      const operation = useOperation()
      const path = new URLPath(operation.path)
      const params = new FunctionParams()
      const withQueryParams = !!schemas.queryParams?.name

      params.add([
        ...getASTParams(schemas.pathParams, {
          typed: true,
          asObject: pathParamsType === 'object',
          override: (item) => ({ ...item, type: `MaybeRef<${item.type}>` }),
        }),
        {
          name: 'params',
          type: schemas.queryParams?.name ? `MaybeRef<${`${factory.name}["queryParams"]`}>` : undefined,
          enabled: !!schemas.queryParams?.name,
          required: isRequired(schemas.queryParams?.schema),
        },
      ])

      const keys = [
        path.toObject({
          type: 'path',
          stringify: true,
          replacer: (pathParam) => `unref(${pathParam})`,
        }),
        withQueryParams ? '...(params ? [params] : [])' : undefined,
      ].filter(Boolean)

      return <Template {...rest} params={params.toString()} keys={keys.join(', ')} />
    }
  },
} as const

type Props = {
  name: string
  typeName: string
  keysFn: (keys: unknown[]) => unknown[]
  factory: {
    name: string
  }
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<FrameworkProps>
}

export function QueryKey({ name, typeName, factory, keysFn, Template = defaultTemplates.react }: Props): ReactNode {
  const {
    options: { pathParamsType },
  } = usePlugin<PluginOptions>()
  const schemas = useOperationSchemas()
  const operation = useOperation()
  const path = new URLPath(operation.path)
  const params = new FunctionParams()
  const withQueryParams = !!schemas.queryParams?.name

  params.add([
    ...getASTParams(schemas.pathParams, {
      typed: true,
      asObject: pathParamsType === 'object',
    }),
    {
      name: 'params',
      type: `${factory.name}["queryParams"]`,
      enabled: !!schemas.queryParams?.name,
      required: isRequired(schemas.queryParams?.schema),
    },
  ])

  const keys = [
    path.toObject({
      type: 'path',
      stringify: true,
    }),
    withQueryParams ? '...(params ? [params] : [])' : undefined,
  ].filter(Boolean)

  return <Template typeName={typeName} name={name} params={params.toString()} keys={keysFn(keys).join(', ')} context={{ factory }} />
}

QueryKey.templates = defaultTemplates
