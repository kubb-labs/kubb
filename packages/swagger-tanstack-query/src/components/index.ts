export { Mutation } from './Mutation.tsx'
export { Operations } from './Operations.tsx'
export { Query } from './Query.tsx'
export { QueryKey } from './QueryKey.tsx'
export { QueryOptions } from './QueryOptions.tsx'
