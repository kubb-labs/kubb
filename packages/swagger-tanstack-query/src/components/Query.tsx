import { PackageManager } from '@kubb/core'
import transformers from '@kubb/core/transformers'
import { FunctionParams, URLPath } from '@kubb/core/utils'
import { Editor, File, Function, usePlugin, useResolveName } from '@kubb/react'
import { pluginKey as swaggerTsPluginKey } from '@kubb/swagger-ts'
import { pluginKey as swaggerZodPluginKey } from '@kubb/swagger-zod'
import { useGetOperationFile, useOperation, useOperationName, useOperationSchemas } from '@kubb/swagger/hooks'
import { getASTParams, getComments, getParams, isRequired } from '@kubb/swagger/utils'

import { getImportNames } from '../utils.ts'
import { QueryImports } from './QueryImports.tsx'
import { QueryKey } from './QueryKey.tsx'
import { QueryOptions } from './QueryOptions.tsx'
import { SchemaType } from './SchemaType.tsx'

import type { ReactNode } from 'react'
import type { QueryOptions as QueryOptionsPluginOptions, Query as QueryPluginOptions } from '../types.ts'
import type { FileMeta, Infinite, PluginOptions, Suspense } from '../types.ts'

type TemplateProps = {
  /**
   * Name of the function
   */
  name: string
  /**
   * Parameters/options/props that need to be used
   */
  params: string
  /**
   * Generics that needs to be added for TypeScript
   */
  generics?: string
  /**
   * ReturnType(see async for adding Promise type)
   */
  returnType?: string
  /**
   * Options for JSdocs
   */
  JSDoc?: {
    comments: string[]
  }
  hook: {
    name: string
    generics?: string
    queryKey: string
    queryOptions: string
  }
  infinite: Infinite | false
}

function Template({ name, generics, returnType, params, JSDoc, hook, infinite }: TemplateProps): ReactNode {
  const isV5 = new PackageManager().isValidSync(/@tanstack/, '>=5')
  const resolvedReturnType = `${returnType} & { queryKey: TQueryKey }`

  if (isV5) {
    return (
      <>
        <Function name={name} export generics={generics} returnType={resolvedReturnType} params={params} JSDoc={JSDoc}>
          {`
         const { query: queryOptions, client: clientOptions = {} } = options ?? {}
         const queryKey = queryOptions?.queryKey ?? ${hook.queryKey}

         const query = ${hook.name}({
          ...${hook.queryOptions} as ${infinite ? 'InfiniteQueryObserverOptions' : 'QueryObserverOptions'},
          queryKey,
          ...queryOptions as unknown as ${infinite ? 'Omit<InfiniteQueryObserverOptions, "queryKey">' : 'Omit<QueryObserverOptions, "queryKey">'}
        }) as ${resolvedReturnType}

        query.queryKey = queryKey as TQueryKey

        return query

         `}
        </Function>
      </>
    )
  }

  return (
    <>
      <Function name={name} export generics={generics} returnType={resolvedReturnType} params={params} JSDoc={JSDoc}>
        {`
       const { query: queryOptions, client: clientOptions = {} } = options ?? {}
       const queryKey = queryOptions?.queryKey ?? ${hook.queryKey}

       const query = ${hook.name}<${hook.generics}>({
        ...${hook.queryOptions},
        queryKey,
        ...queryOptions
      }) as ${resolvedReturnType}

      query.queryKey = queryKey as TQueryKey

      return query

       `}
      </Function>
    </>
  )
}

type FrameworkProps = TemplateProps & {
  context: {
    factory: {
      name: string
    }
    queryKey: string
  }
}

const defaultTemplates = {
  get react() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get solid() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get svelte() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get vue() {
    return function ({ context, hook, ...rest }: FrameworkProps): ReactNode {
      const { factory, queryKey } = context
      const importNames = getImportNames()
      const {
        key: pluginKey,
        options: { pathParamsType },
      } = usePlugin<PluginOptions>()
      const queryOptions = useResolveName({
        name: `${factory.name}QueryOptions`,
        pluginKey,
      })

      const hookName = rest.infinite ? importNames.queryInfinite.vue.hookName : importNames.query.vue.hookName
      const resultType = rest.infinite ? importNames.queryInfinite.vue.resultType : importNames.query.vue.resultType
      const optionsType = rest.infinite ? importNames.queryInfinite.vue.optionsType : importNames.query.vue.optionsType

      const schemas = useOperationSchemas()
      const isV5 = new PackageManager().isValidSync(/@tanstack/, '>=5')
      const params = new FunctionParams()
      const queryParams = new FunctionParams()
      const client = {
        withQueryParams: !!schemas.queryParams?.name,
        withData: !!schemas.request?.name,
        withPathParams: !!schemas.pathParams?.name,
        withHeaders: !!schemas.headerParams?.name,
      }

      const pathParams = getParams(schemas.pathParams, {
        override: (item) => ({
          ...item,
          name: item.name ? `ref${transformers.pascalCase(item.name)}` : undefined,
        }),
      }).toString()

      const resultGenerics = ['TData', `${factory.name}['error']`]

      // only needed for the options to override the useQuery options/params
      const queryOptionsOverrideGenerics = [`${factory.name}['response']`, `${factory.name}['error']`, 'TData', 'TQueryKey']
      const queryOptionsGenerics = ['TData', 'TQueryData']

      params.add([
        ...getASTParams(schemas.pathParams, {
          typed: true,
          asObject: pathParamsType === 'object',
          override: (item) => ({
            ...item,
            name: item.name ? `ref${transformers.pascalCase(item.name)}` : undefined,
          }),
        }),
        {
          name: 'refParams',
          type: `MaybeRef<${schemas.queryParams?.name}>`,
          enabled: client.withQueryParams,
          required: isRequired(schemas.queryParams?.schema),
        },
        {
          name: 'refHeaders',
          type: `MaybeRef<${schemas.headerParams?.name}>`,
          enabled: client.withHeaders,
          required: isRequired(schemas.headerParams?.schema),
        },
        {
          name: 'refData',
          type: `MaybeRef<${schemas.request?.name}>`,
          enabled: client.withData,
          required: isRequired(schemas.request?.schema),
        },
        {
          name: 'options',
          type: `{
              query?: Partial<${optionsType}<${queryOptionsOverrideGenerics.join(', ')}>>,
              client?: ${factory.name}['client']['parameters']
          }`,
          default: '{}',
        },
      ])

      queryParams.add([
        ...getASTParams(schemas.pathParams, {
          typed: false,
          override: (item) => ({
            ...item,
            name: item.name ? `ref${transformers.pascalCase(item.name)}` : undefined,
          }),
        }),
        {
          name: 'refParams',
          enabled: client.withQueryParams,
          required: isRequired(schemas.queryParams?.schema),
        },
        {
          name: 'refHeaders',
          enabled: client.withHeaders,
          required: isRequired(schemas.headerParams?.schema),
        },
        {
          name: 'clientOptions',
          required: false,
        },
      ])

      return (
        <Template
          {...rest}
          params={params.toString()}
          returnType={`${resultType}<${resultGenerics.join(', ')}>`}
          hook={{
            ...hook,
            name: hookName,
            queryOptions: isV5
              ? `${queryOptions}(${queryParams.toString()})`
              : `${queryOptions}<${queryOptionsGenerics.join(', ')}>(${queryParams.toString()})`,
            queryKey: `${queryKey}(${client.withPathParams ? `${pathParams}, ` : ''}${client.withQueryParams ? 'refParams' : ''})`,
          }}
        />
      )
    }
  },
} as const

type Props = {
  factory: {
    name: string
  }
  resultType: string
  hookName: string
  optionsType: string
  infinite: Infinite | false
  query: QueryPluginOptions | false
  queryOptions: QueryOptionsPluginOptions | false
  suspense: Suspense | false
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<FrameworkProps>
  /**
   * This will make it possible to override the default behaviour.
   */
  QueryKeyTemplate?: React.ComponentType<React.ComponentProps<typeof QueryKey.templates.react>>
  /**
   * This will make it possible to override the default behaviour.
   */
  QueryOptionsTemplate?: React.ComponentType<React.ComponentProps<typeof QueryOptions.templates.react>>
}

export function Query({
  factory,
  optionsType,
  hookName,
  resultType,
  Template = defaultTemplates.react,
  QueryKeyTemplate = QueryKey.templates.react,
  QueryOptionsTemplate = QueryOptions.templates.react,
  ...props
}: Props): ReactNode {
  const {
    key: pluginKey,
    options: { dataReturnType, pathParamsType },
  } = usePlugin<PluginOptions>()
  const operation = useOperation()
  const schemas = useOperationSchemas()
  const name = useOperationName({ type: 'function' })
  const isV5 = new PackageManager().isValidSync(/@tanstack/, '>=5')

  const queryKey = useResolveName({
    name: [factory.name, props.infinite ? 'Infinite' : undefined, props.suspense ? 'Suspense' : undefined, 'QueryKey'].filter(Boolean).join(''),
    pluginKey,
  })
  const queryKeyType = useResolveName({
    name: [factory.name, props.infinite ? 'Infinite' : undefined, props.suspense ? 'Suspense' : undefined, 'QueryKey'].filter(Boolean).join(''),
    type: 'type',
    pluginKey,
  })
  const queryOptions = useResolveName({
    name: [factory.name, props.infinite ? 'Infinite' : undefined, props.suspense ? 'Suspense' : undefined, 'QueryOptions'].filter(Boolean).join(''),
    pluginKey,
  })

  const generics = new FunctionParams()
  const params = new FunctionParams()
  const queryParams = new FunctionParams()
  const client = {
    method: operation.method,
    path: new URLPath(operation.path),
    withQueryParams: !!schemas.queryParams?.name,
    withData: !!schemas.request?.name,
    withPathParams: !!schemas.pathParams?.name,
    withHeaders: !!schemas.headerParams?.name,
  }

  generics.add([
    {
      type: 'TData',
      default: props.infinite ? `InfiniteData<${factory.name}["response"]>` : `${factory.name}["response"]`,
    },
    props.suspense ? undefined : { type: 'TQueryData', default: `${factory.name}["response"]` },
    { type: 'TQueryKey extends QueryKey', default: queryKeyType },
  ])

  const pathParams = getParams(schemas.pathParams, {}).toString()
  const resultGenerics = ['TData', `${factory.name}['error']`]
  // only needed for the options to override the useQuery options/params
  // suspense is having 4 generics instead of 5, TQueryData is not needed because data will always be defined
  const queryOptionsOverrideGenerics = props.suspense
    ? [`${factory.name}['response']`, `${factory.name}['error']`, 'TData', 'TQueryKey']
    : [`${factory.name}['response']`, `${factory.name}['error']`, 'TData', 'TQueryData', 'TQueryKey']

  const queryOptionsGenerics = props.suspense ? ['TData'] : ['TData', 'TQueryData']

  params.add([
    ...getASTParams(schemas.pathParams, {
      typed: true,
      asObject: pathParamsType === 'object',
    }),
    {
      name: 'params',
      type: `${factory.name}['queryParams']`,
      enabled: client.withQueryParams,
      required: isRequired(schemas.queryParams?.schema),
    },
    {
      name: 'headers',
      type: `${factory.name}['headerParams']`,
      enabled: client.withHeaders,
      required: isRequired(schemas.headerParams?.schema),
    },
    {
      name: 'data',
      type: `${factory.name}['request']`,
      enabled: client.withData,
      required: isRequired(schemas.request?.schema),
    },
    {
      name: 'options',
      type: `{
    query?: Partial<${optionsType}<${queryOptionsOverrideGenerics.join(', ')}>>,
    client?: ${factory.name}['client']['parameters']
}`,
      default: '{}',
    },
  ])

  queryParams.add([
    ...getASTParams(schemas.pathParams, {
      typed: false,
      asObject: pathParamsType === 'object',
    }),
    {
      name: 'params',
      enabled: client.withQueryParams,
      required: isRequired(schemas.queryParams?.schema),
    },
    {
      name: 'headers',
      enabled: client.withHeaders,
      required: isRequired(schemas.headerParams?.schema),
    },
    {
      name: 'data',
      enabled: client.withData,
      required: isRequired(schemas.request?.schema),
    },
    {
      name: 'clientOptions',
      required: false,
    },
  ])

  const hook = {
    name: hookName,
    generics: [isV5 ? 'any' : `${factory.name}['data']`, `${factory.name}['error']`, 'TData', 'any'].join(', '),
    queryOptions: isV5 ? `${queryOptions}(${queryParams.toString()})` : `${queryOptions}<${queryOptionsGenerics.join(', ')}>(${queryParams.toString()})`,
    queryKey: `${queryKey}(${client.withPathParams ? `${pathParams}, ` : ''}${client.withQueryParams ? 'params' : ''})`,
  }

  return (
    <>
      <QueryKey
        keysFn={props.query ? props.query.queryKey : (keys: unknown[]) => keys}
        Template={QueryKeyTemplate}
        factory={factory}
        name={queryKey}
        typeName={queryKeyType}
      />

      {props.queryOptions && (
        <QueryOptions
          Template={QueryOptionsTemplate}
          factory={factory}
          resultType={optionsType}
          dataReturnType={dataReturnType}
          infinite={props.infinite}
          suspense={props.suspense}
        />
      )}

      {props.query && (
        <Template
          name={[name, props.infinite ? 'Infinite' : undefined, props.suspense ? 'Suspense' : undefined].filter(Boolean).join('')}
          generics={generics.toString()}
          JSDoc={{ comments: getComments(operation) }}
          params={params.toString()}
          returnType={`${resultType}<${resultGenerics.join(', ')}>`}
          hook={hook}
          infinite={props.infinite}
          context={{
            factory,
            queryKey,
          }}
        />
      )}
    </>
  )
}

type FileProps = {
  /**
   * This will make it possible to override the default behaviour.
   */
  templates?: {
    query: typeof defaultTemplates
    queryKey: typeof QueryKey.templates
    queryOptions: typeof QueryOptions.templates
  }
  imports?: typeof QueryImports.templates
}

Query.File = function ({ templates, imports = QueryImports.templates }: FileProps): ReactNode {
  const {
    options: {
      client: { importPath },
      framework,
      infinite,
      suspense,
      query,
      queryOptions,
      parser,
    },
  } = usePlugin<PluginOptions>()

  const schemas = useOperationSchemas()
  const file = useGetOperationFile()
  const fileType = useGetOperationFile({ pluginKey: swaggerTsPluginKey })
  const fileZodSchemas = useGetOperationFile({
    pluginKey: swaggerZodPluginKey,
  })
  const zodResponseName = useResolveName({
    name: schemas.response.name,
    pluginKey: swaggerZodPluginKey,
    type: 'function',
  })

  const factoryName = useOperationName({ type: 'type' })

  const importNames = getImportNames()
  const isV5 = new PackageManager().isValidSync(/@tanstack/, '>=5')
  const Template = templates?.query[framework] || defaultTemplates[framework]
  const QueryOptionsTemplate = templates?.queryOptions[framework] || QueryOptions.templates[framework]
  const QueryKeyTemplate = templates?.queryKey[framework] || QueryKey.templates[framework]
  const Import = imports[framework]

  const factory = {
    name: factoryName,
  }

  return (
    <Editor language="typescript">
      <File<FileMeta> baseName={file.baseName} path={file.path} meta={file.meta}>
        {parser === 'zod' && <File.Import name={[zodResponseName]} root={file.path} path={fileZodSchemas.path} />}
        <File.Import name={'client'} path={importPath} />
        <File.Import name={['ResponseConfig']} path={importPath} isTypeOnly />
        <File.Import
          name={[
            schemas.request?.name,
            schemas.response.name,
            schemas.pathParams?.name,
            schemas.queryParams?.name,
            schemas.headerParams?.name,
            ...(schemas.errors?.map((error) => error.name) || []),
          ].filter(Boolean)}
          root={file.path}
          path={fileType.path}
          isTypeOnly
        />

        <QueryImports Template={Import} isInfinite={false} isSuspense={false} />
        {!!infinite && <QueryImports Template={Import} isInfinite={true} isSuspense={false} />}
        {!!suspense && isV5 && framework === 'react' && <QueryImports Template={Import} isInfinite={false} isSuspense={true} />}
        <File.Source>
          <SchemaType factory={factory} />
          <Query
            factory={factory}
            Template={Template}
            QueryKeyTemplate={QueryKeyTemplate}
            QueryOptionsTemplate={QueryOptionsTemplate}
            infinite={false}
            suspense={false}
            query={query}
            queryOptions={queryOptions}
            hookName={importNames.query[framework].hookName}
            resultType={importNames.query[framework].resultType}
            optionsType={importNames.query[framework].optionsType}
          />
          {!!infinite && (
            <Query
              factory={factory}
              Template={Template}
              QueryKeyTemplate={QueryKeyTemplate}
              QueryOptionsTemplate={QueryOptionsTemplate}
              infinite={infinite}
              suspense={false}
              query={query}
              queryOptions={queryOptions}
              hookName={importNames.queryInfinite[framework].hookName}
              resultType={importNames.queryInfinite[framework].resultType}
              optionsType={importNames.queryInfinite[framework].optionsType}
            />
          )}
          {!!suspense && isV5 && framework === 'react' && (
            <Query
              factory={factory}
              Template={Template}
              QueryKeyTemplate={QueryKeyTemplate}
              QueryOptionsTemplate={QueryOptionsTemplate}
              infinite={false}
              suspense={suspense}
              query={query}
              queryOptions={queryOptions}
              hookName={importNames.querySuspense[framework].hookName}
              resultType={importNames.querySuspense[framework].resultType}
              optionsType={importNames.querySuspense[framework].optionsType}
            />
          )}
        </File.Source>
      </File>
    </Editor>
  )
}

Query.templates = defaultTemplates
