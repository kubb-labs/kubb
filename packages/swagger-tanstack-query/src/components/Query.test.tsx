import { mockedPluginManager } from '@kubb/core/mocks'
import { camelCase, pascalCase } from '@kubb/core/transformers'
import { createRootServer } from '@kubb/react/server'
import { OasManager } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { OperationGenerator } from '../OperationGenerator.tsx'
import { Query } from './Query.tsx'
import { QueryKey } from './QueryKey.tsx'
import { QueryOptions } from './QueryOptions.tsx'

import type { Plugin } from '@kubb/core'
import type { ResolveNameParams } from '@kubb/core'
import type { AppContextProps } from '@kubb/react'
import type { GetOperationGeneratorOptions } from '@kubb/swagger'
import type { PluginOptions } from '../types.ts'

describe('<Query/>', async () => {
  const oas = await OasManager.parseFromConfig({
    root: './',
    output: { path: 'test', clean: true },
    input: { path: 'packages/swagger-tanstack-query/mocks/petStore.yaml' },
  })
  mockedPluginManager.resolveName = ({ type, name }: ResolveNameParams) => {
    if (type === 'file' || type === 'function') {
      return camelCase(`use ${name}`)
    }

    if (type === 'type') {
      return pascalCase(name)
    }
    return name
  }

  const options: GetOperationGeneratorOptions<OperationGenerator> = {
    framework: 'react',
    infinite: false,
    suspense: false,
    dataReturnType: 'data',
    pathParamsType: 'inline',
    templates: {
      query: Query.templates,
      queryKey: QueryKey.templates,
      queryOptions: QueryOptions.templates,
    },
    client: {
      importPath: '@kubb/swagger-client/client',
    },
    parser: undefined,
    query: {
      methods: ['get'],
      queryKey: (key: unknown[]) => key,
    },
    queryOptions: {
      methods: ['get'],
    },
    mutate: false,
  }

  const plugin = { options } as Plugin<PluginOptions>
  const og = await new OperationGenerator(options, {
    oas,
    exclude: [],
    include: undefined,
    pluginManager: mockedPluginManager,
    plugin,
    contentType: undefined,
    override: undefined,
  })

  test('showPetById', async () => {
    const operation = oas.operation('/pets/{uuid}', 'get')
    const context: AppContextProps<PluginOptions['appMeta']> = {
      meta: { pluginManager: mockedPluginManager, plugin },
    }

    const Component = () => {
      return (
        <Oas oas={oas} operations={[operation]} getOperationSchemas={(...props) => og.getSchemas(...props)}>
          <Oas.Operation operation={operation}>
            <Query.File />
          </Oas.Operation>
        </Oas>
      )
    }
    const root = createRootServer({ logger: mockedPluginManager.logger })
    const output = await root.renderToString(<Component />, context)

    expect(output).toMatchFileSnapshot('./__snapshots__/gen/showPetById.ts')
  })
})
