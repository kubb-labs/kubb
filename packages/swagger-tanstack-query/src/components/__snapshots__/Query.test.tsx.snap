// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Query/> > showPetById 1`] = `
"type ShowPetByIdClient = typeof client<ShowPetByIdQueryResponse, never, never>
type ShowPetById = {
  data: ShowPetByIdQueryResponse
  error: never
  request: never
  pathParams: ShowPetByIdPathParams
  queryParams: never
  headerParams: never
  response: ShowPetByIdQueryResponse
  client: {
    parameters: Partial<Parameters<ShowPetByIdClient>[0]>
    return: Awaited<ReturnType<ShowPetByIdClient>>
  }
}

export const ShowPetByIdQueryKey = (petId: ShowPetByIdPathParams['petId'], testId: ShowPetByIdPathParams['testId']) =>
  [{ url: '/pets/:petId', params: { petId: petId } }] as const
export type ShowPetByIdQueryKey = ReturnType<typeof ShowPetByIdQueryKey>
export function ShowPetByIdQueryOptions<TData = ShowPetById['response'], TQueryData = ShowPetById['response']>(
  petId: ShowPetByIdPathParams['petId'],
  testId: ShowPetByIdPathParams['testId'],
  options: ShowPetById['client']['parameters'] = {},
): WithRequired<UseBaseQueryOptions<ShowPetById['response'], ShowPetById['error'], TData, TQueryData>, 'queryKey'> {
  const queryKey = ShowPetByIdQueryKey(petId, testId)

  return {
    queryKey,
    queryFn: async () => {
      const res = await client<ShowPetById['data'], ShowPetById['error']>({
        method: 'get',
        url: \`/pets/\${petId}\`,
        ...options,
      })

      return res.data
    },
  }
}
/**
 * @summary Info for a specific pet
 * @link /pets/:petId */

export function useShowPetById<TData = ShowPetById['response'], TQueryData = ShowPetById['response'], TQueryKey extends QueryKey = ShowPetByIdQueryKey>(
  petId: ShowPetByIdPathParams['petId'],
  testId: ShowPetByIdPathParams['testId'],
  options: {
    query?: Partial<UseBaseQueryOptions<ShowPetById['response'], ShowPetById['error'], TData, TQueryData, TQueryKey>>
    client?: ShowPetById['client']['parameters']
  } = {},
): UseQueryResult<TData, ShowPetById['error']> & { queryKey: TQueryKey } {
  const { query: queryOptions, client: clientOptions = {} } = options ?? {}
  const queryKey = queryOptions?.queryKey ?? ShowPetByIdQueryKey(petId, testId)

  const query = useQuery<ShowPetById['data'], ShowPetById['error'], TData, any>({
    ...ShowPetByIdQueryOptions<TData, TQueryData>(petId, testId, clientOptions),
    queryKey,
    ...queryOptions,
  }) as UseQueryResult<TData, ShowPetById['error']> & { queryKey: TQueryKey }

  query.queryKey = queryKey as TQueryKey

  return query
}
"
`;
